src/messaging/messaging.ts

```diff
@@ -0,0 +1,44 @@
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { FirebaseApp } from 'angularfire2';
+import { messaging } from 'firebase';
+import { requestPermission } from './observable/request-permission';
+import { from } from 'rxjs';
+import { mergeMap } from 'rxjs/operators';
+
+@Injectable()
+export class AngularFireMessaging {
+  messaging: messaging.Messaging;
+  requestPermission: Observable<void>;
+  getToken: Observable<string|null>;
+  tokenChanges: Observable<string|null>;
+  messages: Observable<{}>;
+  requestToken: Observable<string|null>;
+
+  constructor(public app: FirebaseApp) {
+    this.messaging = app.messaging();
+
+    this.requestPermission = requestPermission(this.messaging);
+
+    this.getToken = from(this.messaging.getToken());
+
+    this.tokenChanges = new Observable(subscriber => {
```

Comment:
So now that I'm playing with this I'm not sure I like this API. I think I'd prefer:

`this.tokenChanges() => from(this.messaging.onTokenRefresh);`

`this.token() => this.getToken().pipe(concat(this.tokenChanges())`

Thoughts?